### ISSUE TITLE
Paypal integration

### ISSUE THREAD

**Issue by BrunoSette** [laravel/spark#3]
Stripe is not a worldwide payment company, so It would be really nice to have a Paypal integration as well.

**Comment by davzie** [comment]
I very much doubt this would be a feature Spark would support. I'm guessing it utilizes Cashier quite heavily which again supports Stripe only. May be wrong though ;)

**Comment by NoelDavies** [comment]
Paypal has large fees, disputes are mostly geared towards the customer, and currently, unifying various subscription APIs into one interface has been a failure do the fact that subscription APIs are so different. thephpleague/omnipay tried to do this, but there's just not enough consistency between them all unlike normal payment gateways.

**Comment by tomicakorac** [comment]
As much as I dis-like Paypal, I must say I understand what @BrunoSette is asking. It's not a question whether Paypal is better than Stripe or not (we can all agree on Paypal's bad sides), but rather whether one is able to use Stripe at all or not. Stripe supports a very limited number of countries, and is very slow at adding new ones. I can personally say I realistically expect decades to pass until my country (and my business) is supported by Stripe. So Stripe is simply not an option for a vast number of people who on the other hand would happily use Spark.

If I am ever given a choice, I would always choose 2checkout or Skrill over Paypal though. So I guess my 2 cents would be to ask for an option to customize a payment processing gateway in Spark.

**Comment by tobz-nz** [comment]
I agree with @tomicakorac, another option is kind of a requirement due to Stripe's availability.
I'd prefer to use Stripe, but its just not an option in New Zealand.

**Comment by eduardocruz** [comment]
I'm from Brazil, thus I can't use Stripe either, but maybe this is something to add an issue on the Cashier project instead of Spark. Here is an issue discussing support for Bitcoin for example ... https://github.com/laravel/cashier/issues/186

**Comment by NoelDavies** [comment]
Bitcoin itsself doesn't support subscriptions, and so unless you somehow were able to send out a new QR code for each payment due, before the expiry date then that could work, but would be a massive PITA.

If you want this to work, why not take a look at thephpleague/omnipay#33 and add the support there? Then we can not only allow Spark support for these transactions, but also allow people who use cashier to swap out stripe for omnipay, and also give much of the community an amazing platform for subscriptions as well as one time payments?

**Comment by tomicakorac** [comment]
+1 for the Omnipay. It's an awesome library. It would be a full win if it got supported in Cashier (and thus in Spark too). Although I'm not too big of an optimist, especially since this topic is already closed.

**Comment by NoelDavies** [comment]
I think with the mass interest in Spark and people wanting other gateways, it could push the community to get omnipay started with a subscription api

**Comment by nunomaduro** [comment]
That would be amazing.

**Comment by eduardocruz** [comment]
Source: https://github.com/thephpleague/omnipay

# Recurring Billing

At this stage, automatic recurring payments functionality is out of scope for this library. This is because there is likely far too many differences between how each gateway handles recurring billing profiles. Also in most cases token billing will cover your needs, as you can store a credit card then charge it on whatever schedule you like. Feel free to get in touch if you really think this should be a core feature and worth the effort.

**Comment by NoelDavies** [comment]
@eduardocruz I've seen this, been working with Omnipay for a few months, they say it's out of scope at the moment, but only because there's not enough work behind it. I think with all of the people who want this feature, we could attempt this and give it the backing it's been lacking for a long time.

**Comment by eduardocruz** [comment]
@NoelDavies :+1:  Sounds Great. I went to their Recurring bill issue and saw that you already referenced this thread. Great.

**Comment by aaranmcguire** [comment]
Given omnipay has not even started subscription support, would it be worth putting a proposal in to the cashier project to make it modular like socialite is. (not sure if that would be wanted, but then it allows people to create other drivers) - I've not looked much into the cashier codebase so I don't know how complex it would be to make it modular.

**Comment by NoelDavies** [comment]
@aaranmcguire there's no point in doing this for cashier since Omnipay already has a backing and a structure to go off. It hasn't started subscription support, but it already has the modular support (like socialite) so there'd be no point adding that to cashier when it exists in Omnipay.
